### PR TITLE
Device API and Specification Verification during Registration

### PR DESCRIPTION
**Pull request by twoshark**
### EvaluateSpecificationResponse
- consider http response before attempting to decode device spec

### VerifyDeviceAPI
- add method to ensure device adheres to DC api spec
- hydrates spec information into device

### Specification Check during Registration
- attempt to request and load spec from device after initial health check has passed
- do not activate discovered devices if init, health or spec pull fail
- delay `.Active = true` until spec has successfully been loaded
- add error logging for registration failures

### Local Dev Improvement
- Add db refresh option to make target & script

### CODE REVIEW

File: src/device/specification.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	spec, err := client.EvaluateSpecificationResponse(resp)
+	if err != nil {
+		return err
+	}
+
+	d.LoadFromSpecification(spec)
+	return nil
+}
+
+func (d *Device) LoadFromSpecification(spec Device) {
```
Comment (dtp263) [inline src/device/specification.go:21]: pointer methods dont get mocked which make it harder to test down the line. Id recommend returning itself is a more flexible move. it also then allows for chaining of functions.

```
# before
func (d *Device) LoadFromSpecification(spec Device) {

# after
func (d Device) LoadFromSpecification(spec Device) Device {

File: src/device/specification.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package device
+
+import "fmt"
+
+func (d Device) RequestSpecification(client Client) error {
```
Comment (dtp263) [inline src/device/specification.go:5]: test?

File: src/cluster/registration.go [hunk 3]
```diff
@@ -33,19 +33,39 @@ func (c DeviceCluster) DeviceDiscovery(scanDurationSeconds int) {
...
-			err = d.RunHealthCheck(c.DeviceClient)
-			if err != nil {
-				c.logger.Error(err)
+				break
 			}
 		}
 	}
 }
 
+func (c DeviceCluster) initNewDevice(tmpNewDevice device.NewDeviceParams) error {
```
Comment (dtp263) [inline src/cluster/registration.go:44]: this function does a bunch of stuff. it prolly deserves some in-comments/docs

File: src/device/specification.go (hunk 2 above)
Comment (twoshark) [reply src/device/specification.go]: Added!

File: src/device/specification.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	}
+
+	return spec, nil
+}
+
+func (d Device) LoadFromSpecification(spec Device) Device {
+	d.Description = spec.Description
+	d.Endpoints = spec.Endpoints
+	d.Name = spec.Name
+	return d
```
Comment (dtp263) [inline src/device/specification.go:25]: i think we are missing some assignments here. most importantly `Host` and `Port`. Setting these will enable us to track the same device as it changes IPs and maintains the same MAC address.

File: src/device/specification.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	spec, err := client.EvaluateSpecificationResponse(resp)
+	if err != nil {
+		d.logger.Warn(fmt.Sprintf("Error evaluating specification for [%s]", d.ID.String()))
+		return d, err
+	}
+
+	return spec, nil
+}
+
+func (d Device) LoadFromSpecification(spec Device) Device {
```
Comment (dtp263) [inline src/device/specification.go:21]: also i suppose we can skip for now. but i think this is also where `Endpoints` would be loaded as well.

File: scripts/test.sh [hunk 6]
```diff
@@ -1,8 +1,29 @@
...
-    sed ''/PASS/s//$(printf "\033[32mPASS\033[0m")/'' | \
-    sed ''/FAIL/s//$(printf "\033[31mFAIL\033[0m")/'' | \
-    sed ''/RUN/s//$(printf "\e[96mRUN\e[0m")/'' | \
+refreshdb() {
+    # Stop postgres if already running
+    docker-compose down
+
+    # Clear old volumes
+    docker volume rm devicecommander_postgres || true
+    docker volume rm devicecommander_pgadmin || true
```
[the remark this answers is not in the document]
Comment (twoshark) [reply scripts/test.sh:10]: This was I think meant as more of a dev convenience/break glass procedure. The default is not to tear anything down.

I don't feel strongly about it so I'm happy to remove the option.

[the remark this answers is not in the document]
Comment (dtp263) [reply scripts/test.sh:10]: Meh it's prolly fine. If it helps you dev 🤷🏼‍♂️

**Pull request closed without merging**